src/java.base/share/classes/java/lang/reflect/code/op/CoreOp.java

```diff
@@ -3615,49 +3665,61 @@ public static ConstantOp constant(TypeElement type, Object value) {
     }
 
     /**
-     * Creates an invoke operation.
+     * Creates an invoke instance or class (static) operation.
      *
      * @param invokeDescriptor the invocation descriptor
      * @param args             the invoke parameters
      * @return the invoke operation
      */
     public static InvokeOp invoke(MethodRef invokeDescriptor, Value... args) {
-        return new InvokeOp(invokeDescriptor, List.of(args));
+        return invoke(invokeDescriptor, List.of(args));
     }
 
     /**
-     * Creates an invoke operation.
+     * Creates an invoke instance or class (static) operation.
      *
      * @param invokeDescriptor the invocation descriptor
      * @param args             the invoke parameters
      * @return the invoke operation
      */
     public static InvokeOp invoke(MethodRef invokeDescriptor, List<Value> args) {
-        return new InvokeOp(invokeDescriptor, args);
+        return invoke(invokeDescriptor.type().returnType(), invokeDescriptor, args);
     }
 
     /**
-     * Creates an invoke operation.
+     * Creates an invoke instance or class (static) operation.
      *
      * @param returnType       the invocation return type
      * @param invokeDescriptor the invocation descriptor
      * @param args             the invoke parameters
      * @return the invoke operation
      */
     public static InvokeOp invoke(TypeElement returnType, MethodRef invokeDescriptor, Value... args) {
-        return new InvokeOp(returnType, invokeDescriptor, List.of(args));
+        return invoke(returnType, invokeDescriptor, List.of(args));
     }
 
     /**
-     * Creates an invoke operation.
+     * Creates an invoke instance or class (static) operation.
      *
      * @param returnType       the invocation return type
      * @param invokeDescriptor the invocation descriptor
      * @param args             the invoke parameters
-     * @return the invoke operation
+     * @return the invoke super operation
      */
     public static InvokeOp invoke(TypeElement returnType, MethodRef invokeDescriptor, List<Value> args) {
-        return new InvokeOp(returnType, invokeDescriptor, args);
+        return new InvokeOp.InvokeInstanceClassOp(returnType, invokeDescriptor, args);
+    }
+
+    /**
+     * Creates an invoke super operation.
+     *
+     * @param returnType       the invocation return type
+     * @param invokeDescriptor the invocation descriptor
+     * @param args             the invoke parameters
+     * @return the invoke operation
+     */
+    public static InvokeOp.InvokeSuperOp invokeSuper(TypeElement returnType, MethodRef invokeDescriptor, List<Value> args) {
```

Comment:
The split here hits hard - as you need M * K overloads (in the general case) where M is the number of overloads per kind, and K is the number of kinds we end up supporting.

Reply:
Indeed. Varargs support will also increase the complexity and might motivate a change in the splitting/lumping.

For the majority of the cases so far it just so happens we don't need the kind of invocation and it all works out and one does not need to think too much about it, similar to how `Class::getMethod/getDeclaredMethod` works out. I think it is because of the language rules regarding methods declarations, and all declared methods in a class file are required to have unique name + descriptor regardless of their access flags and attributes ([jvms-4.6](https://docs.oracle.com/javase/specs/jvms/se22/html/jvms-4.html#jvms-4.6))